src/uu/env/src/env.rs

```diff
@@ -165,193 +166,362 @@ pub fn uu_app() -> Command {
-                }
-            } else {
-                break;
-            }
+fn check_and_handle_string_args(
+    arg: &OsString,
+    prefix_to_test: &str,
+    all_args: &mut Vec<std::ffi::OsString>,
+    do_debug_print_args: Option<&Vec<OsString>>,
+) -> UResult<bool> {
+    let native_arg = NCvt::convert(arg);
+    if let Some(remaining_arg) = native_arg.strip_prefix(&*NCvt::convert(prefix_to_test)) {
+        if let Some(input_args) = do_debug_print_args {
+            debug_print_args(input_args); // do it here, such that its also printed when we get an error/panic during parsing
         }
 
-        // read any leftover program arguments
-        for opt in iter {
-            parse_program_opt(&mut opts, opt)?;
+        let arg_strings = parse_args_from_str(remaining_arg)?;
+        all_args.extend(
+            arg_strings
+                .into_iter()
+                .map(from_native_int_representation_owned),
+        );
+
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+#[derive(Default)]
+struct EnvAppData {
+    do_debug_printing: bool,
+    had_string_argument: bool,
+}
+
+impl EnvAppData {
+    fn make_error_no_such_file_or_dir(&self, prog: &OsStr) -> Box<dyn UError> {
+        uucore::show_error!("{}: No such file or directory", prog.quote());
+        if !self.had_string_argument {
+            uucore::show_error!("{}", ERROR_MSG_S_SHEBANG);
         }
+        ExitCode::new(127)
     }
 
-    // GNU env tests this behavior
-    if opts.program.is_empty() && running_directory.is_some() {
-        return Err(UUsageError::new(
-            125,
-            "must specify command with --chdir (-C)".to_string(),
-        ));
+    fn process_all_string_arguments(
+        &mut self,
+        original_args: &Vec<OsString>,
+    ) -> UResult<Vec<std::ffi::OsString>> {
+        let mut all_args: Vec<std::ffi::OsString> = Vec::new();
+        for arg in original_args {
+            match arg {
+                b if check_and_handle_string_args(b, "--split-string", &mut all_args, None)? => {
+                    self.had_string_argument = true;
+                }
+                b if check_and_handle_string_args(b, "-S", &mut all_args, None)? => {
+                    self.had_string_argument = true;
+                }
+                b if check_and_handle_string_args(
+                    b,
+                    "-vS",
+                    &mut all_args,
+                    Some(original_args),
+                )? =>
+                {
+                    self.do_debug_printing = true;
+                    self.had_string_argument = true;
+                }
+                _ => {
+                    all_args.push(arg.clone());
+                }
+            }
+        }
+
+        Ok(all_args)
     }
 
-    // NOTE: we manually set and unset the env vars below rather than using Command::env() to more
-    //       easily handle the case where no command is given
+    #[allow(clippy::cognitive_complexity)]
```

Comment:
for a future PR, it would be nice to address this warning

Reply:
this is moved code. But I'll address it anyway.